Bugfix for wrong timestamps in ros2 bag info

MichaelOrlov opened the pull request:
- This PR addresses the incorrect display of the fractional part of the seconds in the `ros2 bag info` found in the #1744.

Correctly calculate the fractional part for seconds by subtracting `nanoseconds_from_seconds` from `nanoseconds`.

The changes covered by 
https://github.com/ros2/rosbag2/blob/a85672fdf2baf19e6f739879adbbdcfa331948b7/rosbag2_tests/test/rosbag2_tests/test_rosbag2_info_end_to_end.cpp#L41-L61

- Relates #1744

morlov-apexai committed:
Bugfix for wrong timestamps in ros2 bag info

- Correctly calculate fractional part for seconds by subtracting
`nanoseconds_from_seconds` from `nanoseconds`.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Adjust expectations in the "ros2 bag info" integration tests

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

sloretz commented on rosbag2_py/src/rosbag2_py/format_bag_metadata.cpp:
```diff
@@ -42,9 +42,10 @@ std::unordered_map<std::string, std::string> format_duration(
+  auto seconds = std::chrono::duration_cast<std::chrono::seconds>(duration);
+  auto nanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(duration);
+  auto nanoseconds_from_seconds = std::chrono::duration_cast<std::chrono::nanoseconds>(seconds);
+  std::string fractional_seconds = std::to_string((nanoseconds - nanoseconds_from_seconds).count());
```
I think this might still have the bug where leading zeros are lost.

If `duration` is 1 second and 1 nanosecond, `seconds` will equal `1`, `nanoseconds` will equal `1000000001`, and `nanoseconds_from_seconds` will equal `1000000000`.

`nanoseconds - nanoseconds_from_seconds` = `1000000001 - 1000000000` == `1`, and `fractional_seconds` = `std::to_string(1)` which is `"1"`, but the answer we really want is `"000000001"`.

MichaelOrlov replied:
@sloretz Yeah, you are right. I've fixed it by using 
```
fractional_seconds_ss << std::setw(9) << std::setfill('0') << (nanoseconds - nanoseconds_from_seconds).count();
```

MichaelOrlov replied:
Hmm, integration tests failed. Seems need to make adjustments.
I will take a look.

MichaelOrlov replied:
Adjusted expectations in info end-to-end tests by adding leading zero

The real file duration of the recording with sqlite3 is: `70633730` nanoseconds. (from metadata.yaml file)
i.e., regex mask shall be `"0\\.0706.*s"` to match `070633730` nanoseconds. 9 digits for nano.
It was a wrong expectation in tests.

morlov-apexai committed:
Add leading zeros to the fractional seconds in the format_duration(..)

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Adjust expectations in info end-to-end tests by adding leading zero

The real file duration is: 70633730 nanoseconds.
i.e., regex mask shall be "0\\.0706.*s" to match 070633730 nanoseconds.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Pulls: ros2/rosbag2#1745
Gist: https://gist.githubusercontent.com/MichaelOrlov/4e731b24085471856da68644d1675e98/raw/c1397de73131f73c95bb9fcf599ddf9d7208a1c0/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_py rosbag2_tests
TEST args:  --packages-above rosbag2_py rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14261
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21508)](http://ci.ros2.org/job/ci_linux/21508/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15790)](http://ci.ros2.org/job/ci_linux-aarch64/15790/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1004)](http://ci.ros2.org/job/ci_linux-rhel/1004/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22224)](http://ci.ros2.org/job/ci_windows/22224/)

MichaelOrlov commented:
https://github.com/Mergifyio backport jazzy iron humble
